linkerd/app/gateway/src/lib.rs

```diff
@@ -85,161 +93,246 @@ where
+                        .stack
+                        .layer(metrics::StackLabels::inbound("http", "gateway")),
+                )
+                .push(svc::layer::mk(svc::SpawnReady::new))
+                .push(svc::FailFast::layer("Gateway", dispatch_timeout))
+                .push_spawn_buffer(buffer_capacity),
+        )
+        .push_cache(cache_max_idle_age)
+        .push_on_response(
+            svc::layers()
+                .push(http::Retain::layer())
+                .push(http::BoxResponse::layer()),
+        );
 
-    let http_server = {
-        // Cache an HTTP gateway service for each destination and HTTP version.
-        //
-        // The destination is determined from the transport header or, if none was
-        // present, the request URI.
-        //
-        // The client's ID is set as a request extension, as required by the
-        // gateway. This permits gateway services (and profile resolutions) to be
-        // cached per target, shared across clients.
-        let local_id = inbound.runtime().identity.as_ref().map(|l| l.id().clone());
-        let gateway = svc::stack(NewGateway::new(outbound, local_id))
-            .push(profiles::discover::layer(profiles, Allow(allow_discovery)))
-            .instrument(|h: &HttpTarget| debug_span!("gateway", target = %h.target, v = %h.version))
-            .push_on_response(
-                svc::layers()
-                    .push(svc::layer::mk(svc::SpawnReady::new))
-                    .push(
-                        inbound
-                            .runtime()
-                            .metrics
-                            .stack
-                            .layer(metrics::StackLabels::inbound("http", "gateway")),
-                    )
-                    .push(svc::FailFast::layer("Gateway", dispatch_timeout))
-                    .push_spawn_buffer(buffer_capacity),
-            )
-            .push_cache(cache_max_idle_age)
-            .push_on_response(
-                svc::layers()
-                    .push(http::Retain::layer())
-                    .push(http::BoxResponse::layer()),
-            )
-            .push(svc::NewRouter::layer(RouteHttp))
-            .push_http_insert_target::<tls::ClientId>();
-
-        inbound.with_stack(gateway).push_http_server().into_inner()
-    };
-
-    // As gateway connctions are received, dispatch HTTP connections to the HTTP
-    // gateway. If no target was provided AND no protocol, then we attempt HTTP
-    // detection (for legacy clients).
+    // When handling gateway connections from older clients that do not
+    // support the transport header, do protocol detection and route requests
+    // based on each request's URI.
     //
-    // TODO: Handle TCP gateway traffic when a target is set without a protocol.
-    svc::stack(http_server.clone())
-        .push_on_response(svc::MapTargetLayer::new(io::EitherIo::Left))
+    // Non-HTTP connections are refused.
+    let legacy_http = inbound
```

Comment:
should the legacy http stack have its own stack metrics or something? it could be useful to have a way to know how much legacy gateway traffic exists in the mesh.